fluent/migrate/transforms.py

```diff
@@ -92,42 +85,106 @@ def __call__(self, ctx):
+                yield element
             elif isinstance(element, FTL.Expression):
-                flattened.append(FTL.Placeable(element))
+                yield FTL.Placeable(element)
             else:
                 raise RuntimeError(
                     'Expected Pattern, PatternElement or Expression')
-        return flattened
 
     @staticmethod
-    def prune_text_elements(elements):
-        '''Join adjacent TextElements and remove empty ones'''
-        pruned = []
-        # Group elements in contiguous sequences of the same type.
-        for elem_type, elems in itertools.groupby(elements, key=type):
-            if elem_type is FTL.TextElement:
-                # Join adjacent TextElements.
-                text = FTL.TextElement(''.join(elem.value for elem in elems))
-                # And remove empty ones.
-                if len(text.value) > 0:
-                    pruned.append(text)
+    def normalize_text_content(elements):
+        '''Normalize PatternElements with text content.
+
+        Convert TextElements and StringExpressions into TextElements and join
+        adjacent ones.
+        '''
+
+        def get_text(element):
+            if isinstance(element, FTL.TextElement):
+                return element.value
+            elif isinstance(element, FTL.Placeable):
+                if isinstance(element.expression, FTL.StringExpression):
+                    return element.expression.value
+
+        joined = []
+        for current in elements:
+            current_text = get_text(current)
+            if current_text is None:
+                joined.append(current)
+                continue
+
+            previous = joined[-1] if len(joined) else None
+            if isinstance(previous, FTL.TextElement):
+                previous.value += current_text
+            elif len(current_text) > 0:
+                # Normalize to a TextElement
+                joined.append(FTL.TextElement(current_text))
+        return joined
+
+    @staticmethod
+    def preserve_whitespace(elements):
+        # Handle empty values
+        if len(elements) == 0:
+            return [FTL.Placeable(FTL.StringExpression(''))]
+
+        re_leading = re.compile(r'^(?P<whitespace>\s+)(?P<text>.*?)$')
+        re_trailing = re.compile(r'^(?P<text>.*?)(?P<whitespace>\s+)$')
+
+        def extract_whitespace(regex, element):
+            '''Extract leading or trailing whitespace from a TextElement.
+
+            Return a tuple of (Placeable, TextElement) in which the Placeable
+            encodes the extracted whitespace as a StringExpression and the
+            TextElement has the same amount of whitespace removed. The
+            Placeable with the extracted whitespace is always returned first.
+            '''
+            match = re.search(regex, element.value)
+            if match:
+                whitespace = match.group('whitespace')
+                empty_expr = FTL.Placeable(FTL.StringExpression(whitespace))
```

Comment:
Can this be `white_expr`, as it's not empty?

And yes, I still tend to prefer `{""}  foo` to `{"  "}foo`, in which case this would be an `empty_expr` ;-)

Reply:
Or simply `placeable`?